aspnetcore/blazor/file-uploads.md

```diff
@@ -74,21 +77,295 @@ The following example demonstrates multiple image file upload in a component. `I
+
+`UploadResult.cs` in the **`Shared`** project of a hosted Blazor WebAssembly solution:
+
+```csharp
+public class UploadResult
+{
+    public bool Uploaded { get; set; }
+
+    public string FileName { get; set; }
+
+    public string StoredFileName { get; set; }
+
+    public int ErrorCode { get; set; }
+}
+```
+
+`Pages/UploadFiles.razor` in the **`Client`** project of a hosted Blazor solution:
+
+```razor
+@page "/upload-files"
+@using System.IO
+@using System.Linq
+@using System.Net
+@using Microsoft.Extensions.Logging
+@using FileUploadTesting.Shared
+@inject HttpClient Http
+@inject ILogger<UploadFiles> logger
+
+<h1>Upload files</h1>
+
+<p>
+    Upload one or more PNG image files:
+    <InputFile OnChange="@OnInputFileChange" multiple />
+</p>
+
+@if (images.Count > 0)
+{
+    <div class="card">
+        <div class="card-body">
+            @foreach (var image in images)
+            {
+                <figure>
+                    <img class="rounded m-1" src="@image.Thumbnail" />
+                    <figcaption>
+                        @WebUtility.HtmlEncode(image.Name)
+                        <br>
+                        @if (FileUpload(uploadResults, image.Name, logger,
+                            out var result))
+                        {
+                            @result.StoredFileName
+                        }
+                        else
+                        {
+                            <span>
+                                There was an error uploading the file
+                                (Error: @result.ErrorCode).
+                            </span>
+                        }
+                    </figcaption>
+                </figure>
+            }
+        </div>
+    </div>
+}
+
+@code {
+    private IList<Image> images = new List<Image>();
+    private IList<UploadResult> uploadResults =
+        new List<UploadResult>();
+
+    private async Task OnInputFileChange(InputFileChangeEventArgs e)
+    {
+        var format = "image/png";
+        var maxAllowedFiles = 3;
+        var maxWidth = 640;
+        var maxHeight = 480;
+        var thumbnailSize = 100;
+        long maxFileSize = 1024 * 1024 * 15;
+        var upload = false;
+
+        using var content = new MultipartFormDataContent();
+
+        foreach (var imageFile in e.GetMultipleFiles(maxAllowedFiles))
+        {
+            if (uploadResults.SingleOrDefault(
+                f => f.FileName == imageFile.Name) is null)
+            {
+                var image = new Image() { Name = imageFile.Name };
+
+                var resizedImageFile = await imageFile.RequestImageFileAsync(
+                    format, thumbnailSize, thumbnailSize);
+                var buffer = new byte[resizedImageFile.Size];
+                await resizedImageFile.OpenReadStream().ReadAsync(buffer);
+                image.Thumbnail =
+                    $"data:{format};base64,{Convert.ToBase64String(buffer)}";
```

Comment:
This pattern of having large blobs of strings (`Convert.ToBase64String`) perform really poorly with Blazor Server (to the point where the circuit can disconnect). I realize this sits in the WASM section, but users might use this in their server app too. 

Can we skip showing this until we have something [in framework](https://github.com/dotnet/aspnetcore/issues/25274) to help here?

Reply:
Noticed there's a `Convert.ToBase64String` further up too. Are we able to remove that too?

Reply:
FYI @javier

Reply:
Sorry @javier ... pinging @javiercn here.

Reply:
@pranavkm It's covered [here](https://github.com/dotnet/aspnetcore/issues/27885)

Reply:
Also [here](https://github.com/dotnet/aspnetcore/issues/25274)